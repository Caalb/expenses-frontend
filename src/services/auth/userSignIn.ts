import type { AxiosResponse } from 'axios'
import api from '../api'

export interface UserSignInResponse {
  token: string
}

export const userSignIn = async (
  email: string,
  password: string,
): Promise<AxiosResponse<UserSignInResponse>> => {
  const response = await api.post<UserSignInResponse>('/auth/signin', { email, password })

  return response
}
