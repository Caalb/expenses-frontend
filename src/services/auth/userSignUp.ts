import type { AxiosResponse } from 'axios'
import api from '../api'

export interface UserSignUpResponse {
  token: string
}

export const userSignUp = async (
  name: string,
  email: string,
  password: string,
): Promise<AxiosResponse<UserSignUpResponse>> => {
  const response = await api.post<UserSignUpResponse>('/auth/signup', { name, email, password })

  return response
}
