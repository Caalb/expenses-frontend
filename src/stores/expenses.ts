import { createExpense as createExpenseApi } from '@/services/expenses/createExpense'
import { fetchExpenses as fetchExpensesApi } from '@/services/expenses/fetchExpenses'
import type { Expense } from '@/types/expense'
import { defineStore } from 'pinia'
import { ref } from 'vue'

export const useExpensesStore = defineStore('expenses', () => {
  const expenses = ref<Expense[]>([])

  const fetchExpenses = async () => {
    try {
      const response = await fetchExpensesApi()

      expenses.value = response

      return response
    } catch (error) {
      return error
    }
  }

  const createExpense = async (expense: Omit<Expense, 'id'>) => {
    try {
      const response = await createExpenseApi(expense)
      expenses.value = [response.data, ...expenses.value]

      return response
    } catch (error) {
      return error
    }
  }

  return {
    expenses,
    fetchExpenses,
    createExpense,
  }
})
