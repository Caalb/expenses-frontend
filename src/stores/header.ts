import { defineStore } from 'pinia'
import { ref } from 'vue'

export const useHeaderStore = defineStore('header', () => {
  const activeSubmenu = ref<string | null>(null)

  const setActiveSubmenu = (submenu: string) => {
    activeSubmenu.value = activeSubmenu.value === submenu ? null : submenu
  }

  const clearSubmenu = () => {
    activeSubmenu.value = null
  }

  return {
    activeSubmenu,
    setActiveSubmenu,
    clearSubmenu,
  }
})
