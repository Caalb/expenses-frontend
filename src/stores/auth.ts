import { defineStore } from 'pinia'
import { userSignIn } from '../services/auth/userSignIn'
import type { AxiosResponse } from 'axios'
import { userSignUp } from '@/services/auth/userSignUp'

export const useAuthStore = defineStore('auth', () => {
  const handleSignIn = async (
    email: string,
    password: string,
  ): Promise<AxiosResponse | unknown> => {
    try {
      const response = await userSignIn(email, password)
      const { token } = response.data

      localStorage.setItem('token', token)
      return response
    } catch (error: unknown) {
      return error
    }
  }

  const handleSignUp = async (data: Record<string, string>): Promise<AxiosResponse | unknown> => {
    try {
      const response = await userSignUp(data.name, data.email, data.password)
      const { token } = response.data

      localStorage.setItem('token', token)
      return response
    } catch (error: unknown) {
      return error
    }
  }

  return {
    handleSignIn,
    handleSignUp,
  }
})
