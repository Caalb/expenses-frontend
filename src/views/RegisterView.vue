<template>
  <QPage class="flex flex-center">
    <QCard class="register-card">
      <QCardSection class="text-center">
        <h5 class="text-h4 q-mt-none q-mb-md text-grey-8">Registro</h5>
      </QCardSection>

      <QCardSection>
        <Form @submit="onSubmit" class="q-gutter-md" :validation-schema="validationSchema">
          <Field name="name" v-slot="{ errorMessage, value, field }">
            <QInput
              type="text"
              label="Nome"
              outlined
              :model-value="value"
              v-bind="field"
              :error-message="errorMessage"
              :error="!!errorMessage"
            >
              <template #prepend>
                <QIcon name="person" />
              </template>
            </QInput>
          </Field>

          <Field name="email" v-slot="{ errorMessage, value, field }">
            <QInput
              type="email"
              label="E-mail"
              outlined
              :model-value="value"
              v-bind="field"
              :error-message="errorMessage"
              :error="!!errorMessage"
            >
              <template #prepend>
                <QIcon name="email" />
              </template>
            </QInput>
          </Field>

          <Field name="password" v-slot="{ errorMessage, value, field }">
            <QInput
              :type="passwordVisible ? 'text' : 'password'"
              label="Senha"
              outlined
              :model-value="value"
              v-bind="field"
              :error-message="errorMessage"
              :error="!!errorMessage"
            >
              <template #prepend>
                <QIcon name="lock" />
              </template>

              <template #append>
                <QIcon
                  :name="passwordVisible ? 'visibility' : 'visibility_off'"
                  class="cursor-pointer"
                  @click="togglePasswordVisibility"
                />
              </template>
            </QInput>
          </Field>

          <Field name="confirmPassword" v-slot="{ errorMessage, value, field }">
            <QInput
              :type="passwordVisible ? 'text' : 'password'"
              label="Confirmar Senha"
              outlined
              :model-value="value"
              v-bind="field"
              :error-message="errorMessage"
              :error="!!errorMessage"
            >
              <template #prepend>
                <QIcon name="lock" />
              </template>
            </QInput>
          </Field>

          <div class="full-width q-mt-md">
            <QBtn
              type="submit"
              color="primary"
              label="Registrar"
              class="full-width"
              :loading="loading"
            />
          </div>
        </Form>
      </QCardSection>

      <QCardSection class="text-center q-pa-sm">
        <p class="text-grey-6">
          Já tem uma conta?
          <QBtn flat color="primary" label="Faça login" @click="handleLogin" />
        </p>
      </QCardSection>
    </QCard>
  </QPage>
</template>

<script setup lang="ts">
import { ref } from 'vue';
import { QBtn, QCard, QCardSection, QIcon, QInput, QPage, useQuasar } from 'quasar';
import { Form, Field } from 'vee-validate';
import { toTypedSchema } from '@vee-validate/zod';
import { z } from 'zod';
import { useAuthStore } from '@/stores/auth';
import type { AxiosError, AxiosResponse } from 'axios';
import { useRouter } from 'vue-router';
import { errorHandler } from '@/helpers/errorHandler';

const $q = useQuasar()
const router = useRouter()
const { handleSignUp } = useAuthStore()

const validationSchema = toTypedSchema(
  z.object({
    name: z.string()
      .min(3, { message: 'O nome deve ter pelo menos 3 caracteres' })
      .regex(/^[a-zA-Z\s]*$/, { message: 'O nome deve conter apenas letras e espaços' }),
    email: z.string()
      .email({ message: 'E-mail inválido' }),
    password: z.string()
      .min(6, { message: 'A senha deve ter pelo menos 6 caracteres' })
      .regex(/^(?=.*[A-Za-z])(?=.*\d)[A-Za-z\d]{6,}$/, {
        message: 'A senha deve conter pelo menos uma letra e um número'
      }),
    confirmPassword: z.string()
  }).refine((data) => data.password === data.confirmPassword, {
    message: "As senhas não coincidem",
    path: ["confirmPassword"],
  })
)

const loading = ref<boolean>(false)
const passwordVisible = ref<boolean>(false)
const togglePasswordVisibility = (): boolean => passwordVisible.value = !passwordVisible.value

const handleLogin = (): void => {
  router.push({ name: 'login' })
}

const onSubmit = async (values: Record<string, string>): Promise<void> => {
  loading.value = true
  const response = await handleSignUp({
    name: values.name,
    email: values.email,
    password: values.password
  }) as AxiosResponse | AxiosError
  loading.value = false

  if(response && ~~((response.status ?? 0) / 100) === 2) {
    $q.notify({
      color: 'positive',
      message: 'Registro realizado com sucesso!',
      icon: 'check',
    })

    router.push({ name: 'login' })
    return
  }

  errorHandler(response, 'Falha ao realizar registro', $q)
}
</script>

<style scoped>
.register-card {
  width: 100%;
  max-width: 400px;
  padding: 20px;
}
</style>
