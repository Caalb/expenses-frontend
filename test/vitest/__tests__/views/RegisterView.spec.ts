import router from '@/router'
import { useAuthStore } from '@/stores/auth'
import RegisterView from '@/views/RegisterView.vue'
import { createTestingPinia } from '@pinia/testing'
import { installQuasarPlugin } from '@quasar/quasar-app-extension-testing-unit-vitest'
import { mount, VueWrapper } from '@vue/test-utils'
import { Notify } from 'quasar'
import { getByTestId } from 'test/vitest/__utils__'
import { beforeEach, describe, expect, it, vi } from 'vitest'
import type { ComponentPublicInstance } from 'vue'
import { useRouter } from 'vue-router'
import { Form } from 'vee-validate'

installQuasarPlugin({ plugins: { Notify } })

vi.mock('vue-router')

interface RegisterViewInstance extends ComponentPublicInstance {
  onSubmit: () => void
}

const createWrapper = () =>
  mount(RegisterView, {
    global: {
      plugins: [
        createTestingPinia({
          createSpy: vi.fn,
        }),
      ],
      mocks: {
        $router: {
          push: vi.fn(),
        },
      },
    },
  }) as VueWrapper<RegisterViewInstance>

describe('HeaderActions', () => {
  let wrapper: VueWrapper<RegisterViewInstance>
  let store: ReturnType<typeof useAuthStore>

  vi.mocked(useRouter).mockReturnValue({
    ...router,
    push: vi.fn(),
  })

  beforeEach(() => {
    wrapper = createWrapper()
    store = useAuthStore()
  })

  it('should match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should not submit the register form if the form is invalid', async () => {
    const spy = vi.spyOn(wrapper.vm, 'onSubmit')

    await getByTestId(wrapper, 'register-form').trigger('submit')

    expect(spy).not.toHaveBeenCalled()
  })

  it('should submit the register form', async () => {
    await getByTestId(wrapper, 'name-input').setValue('test')
    await getByTestId(wrapper, 'email-input').setValue('test@test.com')
    await getByTestId(wrapper, 'password-input').setValue('123456')
    await getByTestId(wrapper, 'confirmPassword-input').setValue('123456')

    wrapper.findComponent(Form).vm.$emit('submit', {
      name: 'test',
      email: 'test@test.com',
      password: '123456',
      confirmPassword: '123456',
    })

    expect(store.handleSignUp).toHaveBeenCalledWith({
      name: 'test',
      email: 'test@test.com',
      password: '123456',
    })
  })
})
