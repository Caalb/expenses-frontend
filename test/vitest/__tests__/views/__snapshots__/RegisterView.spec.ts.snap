// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HeaderActions > should match snapshot 1`] = `
"<main data-v-bf0653ef="" class="q-page flex flex-center" style="min-height: 768px;">
  <div data-v-bf0653ef="" class="q-card register-card">
    <div data-v-bf0653ef="" class="q-card__section q-card__section--vert text-center">
      <h5 data-v-bf0653ef="" class="text-h4 q-mt-none q-mb-md text-grey-8">Registro</h5>
    </div>
    <div data-v-bf0653ef="" class="q-card__section q-card__section--vert">
      <form data-v-bf0653ef="" novalidate="" class="q-gutter-md" for="register-form" data-testid="register-form"><label data-v-bf0653ef="" class="q-field row no-wrap items-start q-field--outlined q-input q-field--labeled q-field--with-bottom full-width" for="name">
          <!---->
          <div class="q-field__inner relative-position col self-stretch">
            <div class="q-field__control relative-position row no-wrap" tabindex="-1">
              <div class="q-field__prepend q-field__marginal row no-wrap items-center"><i data-v-bf0653ef="" class="q-icon notranslate material-icons" aria-hidden="true" role="presentation">person</i></div>
              <div class="q-field__control-container col relative-position row no-wrap q-anchor--skip"><input class="q-field__native q-placeholder" tabindex="0" aria-label="Nome" name="name" data-testid="name-input" id="name" type="text" value="">
                <div class="q-field__label no-pointer-events absolute ellipsis">Nome</div>
                <!---->
              </div>
            </div>
            <div class="q-field__bottom row items-start q-field__bottom--animated">
              <transition-stub name="q-transition--field-message" appear="false" persisted="false" css="true">
                <div class="q-field__messages col"></div>
              </transition-stub>
              <!---->
            </div>
          </div>
          <!---->
        </label><label data-v-bf0653ef="" class="q-field row no-wrap items-start q-field--outlined q-input q-field--labeled q-field--with-bottom full-width" for="email">
          <!---->
          <div class="q-field__inner relative-position col self-stretch">
            <div class="q-field__control relative-position row no-wrap" tabindex="-1">
              <div class="q-field__prepend q-field__marginal row no-wrap items-center"><i data-v-bf0653ef="" class="q-icon notranslate material-icons" aria-hidden="true" role="presentation">email</i></div>
              <div class="q-field__control-container col relative-position row no-wrap q-anchor--skip"><input class="q-field__native q-placeholder" tabindex="0" aria-label="E-mail" name="email" data-testid="email-input" id="email" type="email" value="">
                <div class="q-field__label no-pointer-events absolute ellipsis">E-mail</div>
                <!---->
              </div>
            </div>
            <div class="q-field__bottom row items-start q-field__bottom--animated">
              <transition-stub name="q-transition--field-message" appear="false" persisted="false" css="true">
                <div class="q-field__messages col"></div>
              </transition-stub>
              <!---->
            </div>
          </div>
          <!---->
        </label><label data-v-bf0653ef="" class="q-field row no-wrap items-start q-field--outlined q-input q-field--labeled q-field--with-bottom full-width" for="password">
          <!---->
          <div class="q-field__inner relative-position col self-stretch">
            <div class="q-field__control relative-position row no-wrap" tabindex="-1">
              <div class="q-field__prepend q-field__marginal row no-wrap items-center"><i data-v-bf0653ef="" class="q-icon notranslate material-icons" aria-hidden="true" role="presentation">lock</i></div>
              <div class="q-field__control-container col relative-position row no-wrap q-anchor--skip"><input class="q-field__native q-placeholder" tabindex="0" aria-label="Senha" name="password" data-testid="password-input" id="password" type="password" value="">
                <div class="q-field__label no-pointer-events absolute ellipsis">Senha</div>
                <!---->
              </div>
              <div class="q-field__append q-field__marginal row no-wrap items-center"><i data-v-bf0653ef="" class="q-icon notranslate material-icons cursor-pointer" aria-hidden="true" role="presentation">visibility_off</i></div>
            </div>
            <div class="q-field__bottom row items-start q-field__bottom--animated">
              <transition-stub name="q-transition--field-message" appear="false" persisted="false" css="true">
                <div class="q-field__messages col"></div>
              </transition-stub>
              <!---->
            </div>
          </div>
          <!---->
        </label><label data-v-bf0653ef="" class="q-field row no-wrap items-start q-field--outlined q-input q-field--labeled q-field--with-bottom full-width" for="confirmPassword">
          <!---->
          <div class="q-field__inner relative-position col self-stretch">
            <div class="q-field__control relative-position row no-wrap" tabindex="-1">
              <div class="q-field__prepend q-field__marginal row no-wrap items-center"><i data-v-bf0653ef="" class="q-icon notranslate material-icons" aria-hidden="true" role="presentation">lock</i></div>
              <div class="q-field__control-container col relative-position row no-wrap q-anchor--skip"><input class="q-field__native q-placeholder" tabindex="0" aria-label="Confirmar Senha" name="confirmPassword" data-testid="confirmPassword-input" id="confirmPassword" type="password" value="">
                <div class="q-field__label no-pointer-events absolute ellipsis">Confirmar Senha</div>
                <!---->
              </div>
            </div>
            <div class="q-field__bottom row items-start q-field__bottom--animated">
              <transition-stub name="q-transition--field-message" appear="false" persisted="false" css="true">
                <div class="q-field__messages col"></div>
              </transition-stub>
              <!---->
            </div>
          </div>
          <!---->
        </label>
        <div data-v-bf0653ef="" class="full-width q-mt-md"><button data-v-bf0653ef="" class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--rectangle bg-primary text-white q-btn--actionable q-focusable q-hoverable full-width" tabindex="0" type="submit"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><span class="block">Registrar</span></span>
            <transition-stub name="q-transition--fade" appear="false" persisted="false" css="true">
              <!---->
            </transition-stub>
          </button></div>
      </form>
    </div>
    <div data-v-bf0653ef="" class="q-card__section q-card__section--vert text-center q-pa-sm">
      <p data-v-bf0653ef="" class="text-grey-6"> Já tem uma conta? <button data-v-bf0653ef="" class="q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--rectangle text-primary q-btn--actionable q-focusable q-hoverable" tabindex="0" type="button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><span class="block">Faça login</span></span></button></p>
    </div>
  </div>
</main>"
`;
